// @salesforce/pwa-kit-runtime/ssr/server/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.X_ORIGINAL_CONTENT_TYPE = exports.X_MOBIFY_QUERYSTRING = exports.X_MOBIFY_FROM_CACHE = exports.X_ENCODED_HEADERS = exports.STRICT_TRANSPORT_SECURITY = exports.STATIC_ASSETS = exports.SLAS_CUSTOM_PROXY_PATH = exports.SET_COOKIE = exports.PROXY_PATH_PREFIX = exports.NO_CACHE = exports.CONTENT_TYPE = exports.CONTENT_SECURITY_POLICY = exports.CONTENT_ENCODING = exports.CACHE_CONTROL = exports.BUILD = exports.APPLICATION_OCTET_STREAM = void 0;
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const APPLICATION_OCTET_STREAM = exports.APPLICATION_OCTET_STREAM = 'application/octet-stream';
const BUILD = exports.BUILD = 'build';
const STATIC_ASSETS = exports.STATIC_ASSETS = 'static_assets';

/**  * @deprecated Use ssr-namespace-paths proxyBasePath instead  */
const PROXY_PATH_PREFIX = exports.PROXY_PATH_PREFIX = '/mobify/proxy';

// All these values MUST be lower case
const CONTENT_TYPE = exports.CONTENT_TYPE = 'content-type';
const CONTENT_ENCODING = exports.CONTENT_ENCODING = 'content-encoding';
const X_ORIGINAL_CONTENT_TYPE = exports.X_ORIGINAL_CONTENT_TYPE = 'x-original-content-type';
const X_MOBIFY_QUERYSTRING = exports.X_MOBIFY_QUERYSTRING = 'x-mobify-querystring';
const X_MOBIFY_FROM_CACHE = exports.X_MOBIFY_FROM_CACHE = 'x-mobify-from-cache';
const X_ENCODED_HEADERS = exports.X_ENCODED_HEADERS = 'x-encoded-headers';
const SET_COOKIE = exports.SET_COOKIE = 'set-cookie';
const CACHE_CONTROL = exports.CACHE_CONTROL = 'cache-control';
const NO_CACHE = exports.NO_CACHE = 'max-age=0, nocache, nostore, must-revalidate';
const CONTENT_SECURITY_POLICY = exports.CONTENT_SECURITY_POLICY = 'content-security-policy';
const STRICT_TRANSPORT_SECURITY = exports.STRICT_TRANSPORT_SECURITY = 'strict-transport-security';

/**  * @deprecated Use ssr-namespace-paths.slasPrivateProxyPath instead  */
const SLAS_CUSTOM_PROXY_PATH = exports.SLAS_CUSTOM_PROXY_PATH = '/mobify/slas/private';
